bvm/ballerina-core/src/main/java/org/ballerinalang/util/TableIterator.java

```diff
@@ -0,0 +1,325 @@
+/*
+*  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
+*
+*  WSO2 Inc. licenses this file to you under the Apache License,
+*  Version 2.0 (the "License"); you may not use this file except
+*  in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*    http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing,
+*  software distributed under the License is distributed on an
+*  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+*  KIND, either express or implied.  See the License for the
+*  specific language governing permissions and limitations
+*  under the License.
+*/
+package org.ballerinalang.util;
+
+import org.ballerinalang.model.ColumnDefinition;
+import org.ballerinalang.model.DataIterator;
+import org.ballerinalang.model.types.BStructType;
+import org.ballerinalang.model.types.BType;
+import org.ballerinalang.model.types.TypeKind;
+import org.ballerinalang.model.types.TypeTags;
+import org.ballerinalang.model.values.BBooleanArray;
+import org.ballerinalang.model.values.BFloatArray;
+import org.ballerinalang.model.values.BIntArray;
+import org.ballerinalang.model.values.BJSON;
+import org.ballerinalang.model.values.BNewArray;
+import org.ballerinalang.model.values.BStringArray;
+import org.ballerinalang.model.values.BStruct;
+import org.ballerinalang.model.values.BXMLItem;
+import org.ballerinalang.util.exceptions.BallerinaException;
+
+import java.sql.Array;
+import java.sql.Blob;
+import java.sql.Connection;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Struct;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Iterator implementation for table data types.
+ *
+ * @since 0.963.0
+ */
+public class TableIterator implements DataIterator {
+
+    protected Connection conn;
+    protected ResultSet rs;
+    protected BStructType type;
+    protected List<ColumnDefinition> columnDefs;
+
+    public TableIterator(Connection conn, ResultSet rs, BStructType type, List<ColumnDefinition> columnDefs) {
+        this.conn = conn;
+        this.rs = rs;
+        this.type = type;
+        this.columnDefs = columnDefs;
+    }
+
+    public TableIterator(Connection conn, ResultSet rs, BStructType type) {
+        this.conn = conn;
+        this.rs = rs;
+        this.type = type;
+        generateColumnDefinitions();
+    }
+
+    @Override
+    public boolean next() {
+        if (rs == null) {
+            return false;
+        }
+        try {
+            return rs.next();
+        } catch (SQLException e) {
+            throw new BallerinaException(e.getMessage(), e);
+        }
+    }
+
+    @Override
+    public void close(boolean isInTransaction) {
+        try {
```

Comment:
shouldn't we have two try catch statements here? say rs.close() fails, then it won't try to close the connection as well. may lead to connection leaks later